metals/src/main/scala/scala/meta/internal/metals/OutlineFilesProvider.scala

```diff
@@ -0,0 +1,191 @@
+import java.util.concurrent.atomic.AtomicBoolean
+import java.{util => ju}
+
+import scala.collection.concurrent.TrieMap
+
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.io.AbsolutePath
+import scala.meta.pc.VirtualFileParams
+import scala.meta.pc.{OutlineFiles => JOutlineFiles}
+
+import ch.epfl.scala.bsp4j.BuildTargetIdentifier
+
+class OutlineFilesProvider(
+    buildTargets: BuildTargets,
+    buffers: Buffers,
+) {
+  private val outlineFiles =
+    new TrieMap[BuildTargetIdentifier, BuildTargetOutlineFilesProvider]()
+
+  def shouldRestartPc(
+      id: BuildTargetIdentifier,
+      reason: PcRestartReason,
+  ): Boolean = {
+    reason match {
+      case DidCompile(true) => true
+      case _ =>
+        outlineFiles.get(id) match {
+          case Some(provider) =>
+            // if it was never compiled successfully by the build server
+            // we don't restart pc not to lose information from outline compile
+            provider.wasSuccessfullyCompiledByBuildServer
+          case None => true
+        }
+    }
+  }
+
+  def onDidCompile(id: BuildTargetIdentifier, wasSuccessful: Boolean): Unit = {
+    outlineFiles.get(id) match {
+      case Some(provider) =>
+        if (wasSuccessful) provider.successfulCompilation()
+      case None =>
+        for {
+          scalaTarget <- buildTargets.scalaTarget(id)
+          // we don't perform outline compilation for Scala 3
+          if (!ScalaVersions.isScala3Version(scalaTarget.scalaVersion))
+        } outlineFiles.putIfAbsent(
+          id,
+          new BuildTargetOutlineFilesProvider(
+            buildTargets,
+            buffers,
+            id,
+            wasSuccessful,
+          ),
+        )
+    }
+  }
+
+  def didChange(id: BuildTargetIdentifier, path: AbsolutePath): Unit = {
+    for {
+      provider <- outlineFiles.get(id)
+    } provider.didChange(path)
+  }
+
+  def getOutlineFiles(
+      buildTargetId: Option[BuildTargetIdentifier]
+  ): Optional[JOutlineFiles] = {
+    val res: Option[JOutlineFiles] =
+      for {
+        id <- buildTargetId
+        provider <- outlineFiles.get(id)
+        outlineFiles <- provider.outlineFiles()
+      } yield outlineFiles
+    res.asJava
+  }
+
+  def enrichWithOutlineFiles(
+      buildTargetId: Option[BuildTargetIdentifier]
+  )(vFile: CompilerVirtualFileParams): CompilerVirtualFileParams = {
+    val optOutlineFiles =
+      for {
+        id <- buildTargetId
+        provider <- outlineFiles.get(id)
+        outlineFiles <- provider.outlineFiles()
+      } yield outlineFiles
+
+    optOutlineFiles
+      .map(outlineFiles => vFile.copy(outlineFiles = Optional.of(outlineFiles)))
+      .getOrElse(vFile)
+  }
+
+  def enrichWithOutlineFiles(
+      path: AbsolutePath
+  )(vFile: CompilerVirtualFileParams): CompilerVirtualFileParams = {
+    val optOutlineFiles =
+      for {
+        bt <- buildTargets.inferBuildTarget(path)
+        provider <- outlineFiles.get(bt)
```

Comment:
```suggestion
        provider <- enrichWithOutlineFiles(bt)
```
? 

The methods seem identical